pkg/federate/kubefed/cluster.go

```diff
@@ -1,9 +1,299 @@
 package kubefed
 
 import (
-	"github.com/submariner-io/admiral/pkg/federate"
+	"fmt"
+	"reflect"
+	"time"
+
+	federate "github.com/submariner-io/admiral/pkg/federate"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/klog"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+	fedv1 "sigs.k8s.io/kubefed/pkg/apis/core/v1beta1"
+	"sigs.k8s.io/kubefed/pkg/client/generic/scheme"
+)
+
+const (
+	tokenKey = "token"
 )
 
 func (f *federator) WatchClusters(handler federate.ClusterEventHandler) error {
-	panic("not implemented")
+	f.startInformerOnce.Do(func() {
+		f.kubeFedClusterInformer.AddEventHandler(f)
+
+		go func() {
+			f.kubeFedClusterInformer.Run(f.stopChan)
+			klog.Info("KubeFedCluster informer stopped")
+		}()
+
+		klog.Info("Started KubeFedCluster informer")
+	})
+
+	var handlerString string
+	if stringer, ok := handler.(fmt.Stringer); ok {
+		handlerString = stringer.String()
+	} else {
+		handlerString = reflect.TypeOf(handler).String()
+	}
+
+	clusterWatcher := &clusterWatcher{
+		eventQueue:           make(chan func(), 1000),
+		enqueueTimerInterval: time.Duration(time.Second * 10),
+		stopChan:             f.stopChan,
+		handler:              handler,
+		handlerString:        handlerString,
+	}
+
+	f.Lock()
+	defer f.Unlock()
+
+	f.clusterWatchers = append(f.clusterWatchers, clusterWatcher)
+
+	go clusterWatcher.eventLoop()
+
+	for clusterID, kubeConfig := range f.clusterMap {
+		clusterWatcher.onAdd(clusterID, kubeConfig)
+	}
+
+	klog.Infof("Federator added ClusterEventHandler \"%s\"", handlerString)
+
+	return nil
+}
+
+func (f *federator) OnAdd(obj interface{}) {
+	klog.V(3).Infof("In federated cluster watcher OnAdd for %#v", obj)
+
+	kubeFedCluster := obj.(*fedv1.KubeFedCluster)
+
+	kubeConfig, err := f.buildFederatedClusterConfig(kubeFedCluster)
+	if err != nil {
+		klog.Errorf("Error building kube config for federated cluster %#v: %v", kubeFedCluster, err)
+		return
+	}
+
+	f.Lock()
```

Comment:
Code needs to be safe above performant.

I wonder if we can just unlock before calling the watchers OnAdd.

Reply:
> (non-voting) for now I think it's okey to have this lock but I envision it being a future problem impacting performance.
> 
> If I understand correctly you have it to avoid a race between the onAdd and the onDelete, no?

It's needed for memory synchronization between multiple threads and also for atomicity wrt `WatchClusters`.

Reply:
> Code needs to be safe above performant.
> 
> I wonder if we can just unlock before calling the watchers OnAdd.

We do need atomicity wrt `OnAdd` as well as it enqueues data on the handler's channel and we need to ensure proper ordering of the handler's event stream.  That said, technically it shouldn't matter either way in this case as we don't expect the informer to notify concurrently. However we still need to synchronize access to `clusterWatchers` wrt `WatchClusters` so we'd have to make a copy of `clusterWatchers` before exiting the critical section to be safe, but I don't think that's needed.

Reply:
> Code needs to be safe above performant.

sure. this comment was done before we talked about this.

Said that, I think this is not the language's intent. The language favors a chan/go-routine approach. 

Thus I believe it should be best done that way to allow the language (and the compiler) optimize the binary. That way is both performant and safe.

As an alternative, kubernetes does have an abstraction that suits these needs (workqueue) and I would favor using that if the channel/go-routine approach is not favored. This last pattern is well documented for writing controllers: [[1]](https://github.com/kubernetes/sample-controller/blob/master/controller.go) [[2]](https://itnext.io/how-to-create-a-kubernetes-custom-controller-using-client-go-f36a7a7536cc) [[3]](https://medium.com/speechmatics/how-to-write-kubernetes-custom-controllers-in-go-8014c4a04235) [[4]](https://engineering.bitnami.com/articles/kubewatch-an-example-of-kubernetes-custom-controller.html)

Reply:
I tried using the kubernetes workqueue mechanism but did not find it suitable here and also overly complex. Also it does not obviate the need for synchronization as we still need to access the shared memory in `WatchClusters` (and possibly other places as well as the code evolves). 

While Go encourages the uses on channels over shared memory as a guideline to alleviate races etc, I think shared memory with  a simple mutex is suitable here.